### PR TITLE
Provide access to command line arguments within top-level statements

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to #41704.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @RikkiGibson, @dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Binder/InMethodBinder.cs [hunk 1]
```diff
@@ -235,9 +235,13 @@ protected override void AddLookupSymbolsInfoInSingleBinder(LookupSymbolsInfo res
 
         private static bool ReportConflictWithParameter(Symbol parameter, Symbol newSymbol, string name, Location newLocation, DiagnosticBag diagnostics)
         {
-            var oldLocation = parameter.Locations[0];
+#if DEBUG
+            var locations = parameter.Locations;
+            Debug.Assert(!locations.IsEmpty || parameter.IsImplicitlyDeclared);
+            var oldLocation = locations.IsEmpty ? Location.None : locations[0];
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/InMethodBinder.cs:241]: consider using `locations.FirstOrNone()`

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 2]
```diff
@@ -3688,6 +3730,333 @@ void M()
...
+            var symbols = model.LookupSymbols(nameRef.SpanStart);
+            Assert.Contains(testType, symbols);
+            Assert.False(symbols.Any(s => s.Kind == SymbolKind.Parameter));
+            Assert.Same(testType, model.LookupSymbols(nameRef.SpanStart, name: "args").Single());
+
+            nameRef = nameRefs[1];
+            Assert.Equal("System.Console.WriteLine(args)", nameRef.Parent.Parent.Parent.ToString());
+            var parameter = model.GetSymbolInfo(nameRef).Symbol;
+            Assert.Equal("System.String[] args", parameter.ToTestDisplayString());
+            Assert.Equal("<simple-program-entry-point>", parameter.ContainingSymbol.ToTestDisplayString());
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:3824]: >simple-program [](start = 27, length = 14)

Should this name be changed to align with the feature name? Better yet, could we report "$Main()" instead?

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 3]
```diff
@@ -4033,6 +4402,101 @@ void local()
...
+    {
+        System.Console.WriteLine(x);
+    }
+}
+";
+
+            var comp = CreateCompilation(new[] { text1 }, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+
+            CompileAndVerify(comp, expectedOutput: "4");
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:4498]: >} [](start = 8, length = 1)

Please test use of `args` inside a local function.
```C#
local();
static object local() => args;
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:3824]: > Should this name be changed to align with the feature name?

I am planning to make this adjustment separately. It is not primarily related to the goal of this PR. 

---
In reply to: [424087858](https://github.com/dotnet/roslyn/pull/44167#discussion_r424087858) [](ancestors = 424087858)

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 4]
```diff
@@ -989,6 +1000,31 @@ public void LocalDeclarationStatement_13()
...
+                );
+
+            Assert.False(comp.NullableSemanticAnalysisEnabled); // To make sure we test incremental binding for SemanticModel
+
+            var tree1 = comp.SyntaxTrees[0];
+            var model1 = comp.GetSemanticModel(tree1);
+            var symbol1 = model1.GetDeclaredSymbol(tree1.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().First());
+            Assert.Equal("System.String args", symbol1.ToTestDisplayString());
+            Assert.Same(symbol1, model1.GetSymbolInfo(tree1.GetRoot().DescendantNodes().OfType<IdentifierNameSyntax>().Where(id => id.Identifier.ValueText == "args").Single()).Symbol);
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:1026]: >} [](start = 8, length = 1)

Consider testing a range variable.
```
_ = from args in new object[0] select args;
```

**Pull request merged**